dask/dataframe/io/parquet.py

```diff
@@ -580,24 +585,56 @@ def _read_pyarrow(fs, fs_token, paths, columns=None, filters=None,
     return out_type(task_plan, task_name, meta, divisions)
 
 
-def _get_pyarrow_dtypes(schema):
+def _get_pyarrow_dtypes(schema, categories):
     dtypes = {}
     for i in range(len(schema)):
         field = schema[i]
         numpy_dtype = field.type.to_pandas_dtype()
         dtypes[field.name] = numpy_dtype
 
+    if categories:
+        for cat in categories:
+            dtypes[cat] = "category"
+
     return dtypes
 
 
 def _read_pyarrow_parquet_piece(fs, piece, columns, index_cols, is_series,
-                                partitions):
+                                partitions, categories):
+    import pyarrow as pa
+
     with fs.open(piece.path, mode='rb') as f:
         table = piece.read(columns=index_cols + columns,
                            partitions=partitions,
                            use_pandas_metadata=True,
                            file=f)
-    df = table.to_pandas()
+
+    if pa.__version__ < distutils.version.LooseVersion('0.8.0'):
+        df = table.to_pandas()
+        for cat in categories:
+            df[cat] = df[cat].astype('category')
+    elif pa.__version__ < distutils.version.LooseVersion('0.9.0'):
+        # PyArrow can only reliably dictionary encode array that consist of
+        # a single chunk. The necessary C++ functionality for a whole column is
+        # missing from the Python interface.
+        remaining_cats = []
+        for i in range(table.num_columns):
+            col = table.column(i)
+            if col.name in categories:
+                new_col = col
+                if col.data.num_chunks != 1:
+                    remaining_cats.append(col.name)
+                else:
+                    array = col.data.chunk(0)
+                    array = array.dictionary_encode()
+                    new_col = pa.column(col.name, array)
+                    table = table.remove_column(i)
+                    table = table.add_column(i, new_col)
+        df = table.to_pandas()
+        for cat in remaining_cats:
+            df[cat] = df[cat].astype('category')
```

Comment:
Is there a strong benefit to doing it this way? I'm half tempted to just fall back to the simple pandas method due to maintenance costs. I suspect people using pyarrow are fairly quick to upgrade to new releases.

Reply:
Given that we will hopefully release pyarrow in the next two weeks, I'm going to remove this again